File: PhotonCore/src/main/java/com/outoftheboxrobotics/photoncore/hardware/i2c/commands/PhotonLynxI2cWriteStatusQueryCommand.java
package com.outoftheboxrobotics.photoncore.hardware.i2c.commands;

import com.outoftheboxrobotics.photoncore.PhotonCommandBase;
import com.qualcomm.hardware.lynx.LynxModuleIntf;
import com.qualcomm.hardware.lynx.LynxNackException;
import com.qualcomm.hardware.lynx.commands.LynxMessage;
import com.qualcomm.hardware.lynx.commands.core.LynxI2cWriteStatusQueryCommand;
import com.qualcomm.hardware.lynx.commands.standard.LynxAck;
import com.qualcomm.hardware.lynx.commands.standard.LynxNack;

import java.util.concurrent.CompletableFuture;

public class PhotonLynxI2cWriteStatusQueryCommand extends LynxI2cWriteStatusQueryCommand implements PhotonCommandBase {
    private CompletableFuture<LynxMessage> future;
    public PhotonLynxI2cWriteStatusQueryCommand(LynxModuleIntf module) {
        super(module);
    }

    public PhotonLynxI2cWriteStatusQueryCommand(LynxModuleIntf module, int busZ) {
        super(module, busZ);
    }

    @Override
    public void onResponseReceived(LynxMessage response) {
        super.onResponseReceived(response);
        future.complete(response);
    }

    @Override
    public void onAckReceived(LynxAck ack) {
        super.onAckReceived(ack);
        future.complete(ack);
    }

    @Override
    public void onNackReceived(LynxNack nack) {
        super.onNackReceived(nack);
        future.complete(nack);
    }

    @Override
    public CompletableFuture<LynxMessage> getResponse() throws LynxNackException {
        return future;
    }
}

File: PhotonCore/src/main/java/com/outoftheboxrobotics/photoncore/HAL/I2C/Commands/PhotonLynxI2cWriteStatusQueryCommand.kt
package com.outoftheboxrobotics.photoncore.HAL.I2C.Commands

import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonCommandBase
import com.qualcomm.hardware.lynx.LynxModuleIntf
import com.qualcomm.hardware.lynx.LynxNackException
import com.qualcomm.hardware.lynx.commands.LynxMessage
import com.qualcomm.hardware.lynx.commands.core.LynxI2cWriteStatusQueryCommand
import com.qualcomm.hardware.lynx.commands.standard.LynxAck
import com.qualcomm.hardware.lynx.commands.standard.LynxNack
import java.util.concurrent.CompletableFuture

class PhotonLynxI2cWriteStatusQueryCommand : LynxI2cWriteStatusQueryCommand, PhotonCommandBase {
    private val future: CompletableFuture<LynxMessage> = CompletableFuture();

    constructor(module: LynxModuleIntf?) : super(module)
    constructor(module: LynxModuleIntf?, busZ: Int) : super(module, busZ)

    override fun onResponseReceived(response: LynxMessage) {
        super.onResponseReceived(response)
        future.complete(response)
    }

    override fun onAckReceived(ack: LynxAck) {
        super.onAckReceived(ack)
        future.complete(ack)
    }

    override fun onNackReceived(nack: LynxNack) {
        super.onNackReceived(nack)
        future.complete(nack)
    }

    override fun getResponse(): CompletableFuture<LynxMessage> {
        return future
    }
}